[IV-21-22] Objetivo 7

sgIOlas opened the pull request:
Se realizan los contenidos del objetivo. Ahora, la aplicación tiene soporte para:
- Registrar una serie de eventos predefinidos de forma sencilla.
- Obtener parámetros de configuración como el directorio de logs desde etcd.
- Hacer uso de variables de entorno desde un fichero .env para parámetros potencialmente sensibles como credenciales de acceso al servidor de etcd.

Todo esto se ha realizado teniendo en cuenta las mejores prácticas tanto para logging (Interfaz estándar de logging, logs formateados, inyección de dependencias, escritura a archivo, etc.) como para configuración remota (.env en .gitignore o autenticación para el servidor de etcd).

De estar todo correcto, Closes #47 #48 #49 #50.

sgIOlas committed:
Refactoriza algunos identificadores

Seguir reglas de estilo de Go:
- Algo que no se planea que sea exportado debería de empezar por minúscula.
- No usar _ para identificadores.

sgIOlas committed:
Añadir dependencia logrus #49

Se actualizan go.mod y go.sum mediante "go mod tidy" para reflejar la nueva dependencia.

sgIOlas committed:
Avance inicial #49

Primer acercamiento a un wrapper básico para logs.

sgIOlas committed:
Pequeñas correcciones de tests #49

sgIOlas committed:
Interfaz más independiente #49

Se pretende separar lo máximo posible el uso de la dependencia de logrus de nuestra interfaz de logging. Añadir además un campo para registrar la versión de Go en los logs.

sgIOlas committed:
Nuevo sistema de logging #49

Cambios en el wrapper para facilitar el registro de eventos mediante funciones asignadas a cada evento. Además, se adapta  la forma "idiomática" de Go de retorno de nil en caso de que todo haya salido bien.

sgIOlas committed:
Logging y bound checks #49

Se completa la implementación básica de la interfaz de logging y se añaden comprobaciones para comprobar y ajustar los parámetros de una banda.

Closes #49.

sgIOlas committed:
Avances conf. remota #50

Se añade la dependencia de etcd al proyecto además del paquete config. En este paquete se implementa la conexión con el servidor de almacenamiento distribuido y se realiza la lógica para obtener el path de los logs primero desde etcd y luego desde una variable de entorno mediante godotenv.

sgIOlas committed:
Añadir .env a .gitignore #50

Evitar subir al repositorio el fichero que contiene las variables de entorno, ya que podría almacenar información sensible.

sgIOlas committed:
Unificar paquete config #50

Separar el paquete no tiene sentido por el momento ya que se utilizan el uno al otro. Separar los paquetes da lugar a problemas de bucles de imports.

sgIOlas committed:
Soporte para logs en archivo #49

Se incluye la función prepareLogFile para gestionar la apertura/creación del fichero de logs. Además, se añaden dos nuevos eventos a registrar.

sgIOlas committed:
Añadir timeout conex. etcd #50

Previamente, si el servidor de etcd no estaba disponible, el programa detenía su ejecución indefinidamente. Añadir un timeout de dos segundos al contexto para evitarlo.

sgIOlas committed:
Cambio de paquete de logging a zerolog #49

Se había tomado una mala decisión previamente eligiendo a logrus como el sistema de logs del proyecto, ya que como se indica en su [repo](https://github.com/sirupsen/logrus) el proyecto ha quedado en modo mantenimiento debido a que les resulta muy difícil realizar cambios sin romper compatibilidad. Es por esto que cambiamos nuestra dependencia de logging a una que ellos mismos recomiendan: zerolog.

sgIOlas committed:
Añadir timestamp por defecto al logger #49

De esta forma no es necesario que cada generación de un log tenga que incluir explícitamente en el código el timestamp.

sgIOlas committed:
Añadir justificación técnica #47

sgIOlas committed:
Justificación técnica c. remota #48

[no ci]

sgIOlas committed:
Completar documentación #48

sgIOlas committed:
Usar autenticación para etcd #50

Para mayor seguridad, se hará uso de un log in básico de los clientes en el servidor de etcd. Las credenciales se obtendrán de variables de entorno.

sgIOlas committed:
Añadir clave iv.yaml #51

Además se corrige la falta de índice para la entrada de configuración remota de la documentación.

[no ci]

Olasergiolas commented:
⛹ Revisores → @alexespana @modejota @joaquingv12 por si queréis ir echándole un vistazo.

JJ commented:
⛹ Revisores → @alexespana @modejota @joaquingv12

sgIOlas committed:
Más campos por defecto #49

Añadir también el campo Fields al Logger que generamos para evitar repetir su inclusión.

sgIOlas committed:
Gestionar error cnx etcd #50

Si el error que devuelve la creación del cliente no se gestiona, la función searchLogPath() genera un segfault.

modejota commented:
Es verdad que no he trabajado nunca con Go, pero yo lo veo todo en orden. 
Me gusta el detalle de haber testeado varias opciones y haberlo reflejado en la documentación, así como me resulta curioso que el paquete por defecto de logging de Go sea tan "pobre".

alexespana reviewed (commented):
Nunca he trabajado con go, por lo que en ese aspecto no puedo ser objetivo, sin embargo, las justificaciones técnicas las veo muy bien desarrolladas :+1:

Olasergiolas commented:
Creo que el PR ya está listo para revisión @JJ

JJ commented on pkg/autoeq/config.go:
```diff
@@ -0,0 +1,64 @@
+		defer cli.Close()
+	}
+
+	NewLogger().ConfigInfoLog(logPathEnvVar, currentLogPath)
```
*Separation of concerns*: el módulo debe sólo devolver una estructura de datos con la configuración, no configurar.

Olasergiolas replied:
Intentaré darle otra vuelta a la implementación.

JJ commented on pkg/autoeq/equalizer.go:
```diff
@@ -1,9 +1,9 @@
-	DefaultEqMode        string  = "IIR"
-	DefaultSplitChannels bool    = false
-	DefaultOutputGain    float32 = 0.0
+	defaultEqMode        string  = "IIR"
```
¿Y estos cambios a minúscula?

Olasergiolas replied:
En Go se utiliza la primera letra en mayúscula de un identificador cuando se quiere poder exportar a otros paquetes y dado que el uso que se le va a dar a esas constantes es interno, no nos interesa exportarlas.

JJ commented on pkg/autoeq/log_wrapper.go:
```diff
@@ -0,0 +1,89 @@
+	return f, nil
+}
+
+func (l *MyLogger) SuccessLog(context string) {
```
Buen wrapping específico de la aplicación.

JJ commented on pkg/autoeq/output.go:
```diff
@@ -21,7 +21,12 @@ func NewOutput(eq equalizer) outputWrapper {
+	lines, err := OpenParametricData(configPath)
+
+	if err != nil {
+		NewLogger().FileNotOpenedLog(configPath)
```
¿Creas un logger cada vez? ¿No compartes el objeto?

Olasergiolas replied:
Me planteé aplicar singleton pero no tenía muy claro si era lo más correcto. Sería preferible ¿Verdad?

JJ reviewed (changes requested):
Está casi perfecto, pero
* No testeas ni el objeto de log ni el de configuración. Lo del primero no es tan importante (aunque es conveniente, claro), el segundo sí.
* Deberías tratar de "inyectar" un objeto de log en vez de generarlo donde lo usas. Y también lo usas bastante poco...
* Asegúrate de que el milestone describe correctamente esta funcionalidad como parte del PMV.

sgIOlas committed:
Reusa instancia de logging #51

Aplicar Singleton para no tener que estar creando un logger con tanta frecuencia.

sgIOlas committed:
Registro de más eventos #51

Se añade un nuevo evento para registrar paso de parámetros inesperados. También se hace uso del logger en más situaciones.

sgIOlas committed:
Separación de intereses config #51

Separamos las tareas de generar una configuración y configurar la aplicación. Ahora, como se ha sugerido en la discusión del PR #51, en config se retorna una estructura de datos con la configuración en lugar de configurar directamente.

sgIOlas committed:
Testear logs y config. remota #51

Previamente, el correcto funcionamiento de la nueva funcionalidad introducida para este objetivo no era testeado. Introducir dos nuevos test:
- TestLogWrapper: Comprueba que el mensaje de la última entrada en los logs es el esperado.
- TestGetConfig: Comprueba que el parámetro de configuración "logPath" se fija al valor esperado.

sgIOlas committed:
Ajustar logpath desde task #51

Por algún motivo, la solución anterior solo funcionaba si los tests se ejecutaban desde VSCode. Fijar la env-var desde task para que funcione correctamente.

sgIOlas committed:
Unificar uso de tests #51

Se añade la action indicada en la documentación oficial de Task y se hace uso del task runner para ejecutar los tests.

Olasergiolas commented:
He intentado tratar todo lo comentado en tu revisión @JJ.

Listo para revisión.

JJ reviewed (approved):
:+1:
